boards: arm: nucleo_f756zg: Created new board nucleo_f756zg

ajpcode opened the pull request:
Created board support for the nucleo_f756zg.

Features supported include:
 - GPIO
 - UART
 - ETH
 - USB
 - I2C
 - PWM
- SPI

dbkinder commented on boards/arm/nucleo_f756zg/doc/nucleof756zg.rst:
```diff
@@ -0,0 +1,216 @@
+    - 2 user and reset push-buttons
+    - 32.768 kHz crystal oscillator
+
+    Board connectors:
```
This should be a list item, with a nested list, and the remaining items should be list items too:

    - Board connectors:

      - USB with Micro-AB
      - SWD
      - Ethernet RJ45 (depending on STM32 support)
      - ST Zio connector including Arduino* Uno V3
      - ST morpho

    - Flexible power-supply options: ST-LINK USB VBUS or external sources.
    - On-board ST-LINK/V2-1 debugger/programmer with USB re-enumeration
      capability: mass storage, virtual COM port and debug port.
    - Comprehensive free software libraries and examples available with the
      STM32Cube MCU package.
    - Arm* Mbed Enabled* compliant (only for some Nucleo part numbers)

dbkinder commented on boards/arm/nucleo_f756zg/doc/nucleof756zg.rst:
```diff
@@ -0,0 +1,216 @@
+The Nucleo F756ZG board features a ST Zio connector (extended Arduino Uno V3)
+and a ST morpho connector. Board is configured as follows:
+
+    - UART_2 TX/RX/RTS/CTS : PD5/PD6/PD4/PD3
```
Don't indent this list

ajpcode commented:
Latest set of commits address documentation changes requested by @dbkinder and adds SPI support.

dbkinder reviewed (approved):
+1 for docs, thanks!

ajpcode commented:
Latest commit rebased to pull the f7 pinmux definitions from now merged PR #10419.

ajpcode commented:
Added default selection of ports gpio-f and gpio-g on selection of the gpio device driver for the stm32f756xg SOC family.

erwango commented on boards/arm/nucleo_f756zg/pinmux.c:
```diff
@@ -0,0 +1,43 @@
+
+/* NUCLEO-F756ZG pin configurations  */
+static const struct pin_config pinconf[] = {
+#ifdef CONFIG_UART_STM32_PORT_2
```
Why configuring this one, if not used?

ajpcode replied:
Enabled in .dts and device selected by default on CONFIG_SERIAL. 

(I'm using this uart for an offline project. I had overlooked the addition to the branch, this is why the pin definition was present).

erwango commented on dts/arm/st/f7/stm32f756Xg.dtsi:
```diff
@@ -0,0 +1,10 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <st/f7/stm32f746Xg.dtsi>
```
Even if this works for now, this is not correct. A package should include a chip, not another SoC's package.
So here, we should have:
```
#include <st/f7/stm32f756Xg.dtsi>
```
And, in st/f7/stm32f756Xg.dtsi:
```
#include <st/f7/stm32f746.dtsi>
```

ajpcode replied:
OK, thanks.
I've based the soc section on the 769 - hopefully this corrects this.

erwango commented on boards/arm/nucleo_f756zg/doc/nucleof756zg.rst:
```diff
@@ -138,6 +140,10 @@ and a ST morpho connector. Board is configured as follows:
 - PWM : PE13
+- SPI : PA4, PA5, PA6, PA7
+
+Note. The Arduino Uno v3 specified SPI device conflicts with the on-board ETH
```
Interesting node. I wonder if there is a way to generate a warning if this is enabled (Using Kconfig maybe?)

ajpcode replied:
I did wonder initially about adding a #warning to the source to generate a compiler warning on compilation of pinmux.c if pins are conflicted.

I also considered kconfig dependency configuration but (for example) the spi_1 mosi signal can obviously be muxed to another pin for spi_1 to still work correctly with the on board Ethernet (although would break Arduino compatibility by doing so).

The documentation route seemed to be least messy/conflicting but happy to go with any suggestions you may have on how best to deal with this.

erwango replied:
Had no ideas on mind than the ones you proposed. 
Hence, it seems documentation is the best option we have, even if I'm certain people will face the issue before reading the doc.
Can you add comments as well in dts (before each conflicting node: arduino, ethernet) and pinmux? This will maximize the chances people read some info in this conflict.

ajpcode replied:
Absolutely, I'll get that sorted. When I get a chance I'll also stand up an equivalent PR for the f746 nucleo comments as it is also affected by this feature.

erwango reviewed (changes requested):
Few comments, otherwise looks good

Commit:
boards: arm: nucleo_f756zg: Created new board and soc definition

Added board definitions for nucleo_f756zg. Features include gpio,
pinmux, uart (ST Zio, ST-Link and Arduino Uno v3 interfaces).

Added basic documentation and some soc definitions for the
stm32 f756XX soc.

Signed-off-by: AJ Palmer <ajpcode@hotmail.com>

ajpcode commented:
Latest commit rebased on master and makes changes requested by @erwango .

galak commented on boards/arm/nucleo_f756zg/board.h:
```diff
@@ -0,0 +1,13 @@
+#ifndef __INC_BOARD_H
+#define __INC_BOARD_H
+
+#include <soc.h>
```
Remove soc.h from here.

ajpcode replied:
As requested, removed.

Commit:
boards: arm: nucleo_f756zg: Added ethernet support

Added pinmux and defconfig information required to use the on-board
ethernet.

Signed-off-by: AJ Palmer <ajpcode@hotmail.com>

Commit:
boards: arm: nucleo_f756zg: Added USB support

Added pinmux and device tree information required to use the on-board
usb otg port.

Signed-off-by: AJ Palmer <ajpcode@hotmail.com>

Commit:
boards: arm: nucleo_f756zg: Added I2C support

Added I2C support to the I2C bus connected to the arduino uno v3
connector.

Signed-off-by: AJ Palmer <ajpcode@hotmail.com>

Commit:
boards: arm: nucleo_f756zg: Added PWM support

Added PWM support for one of the gpios connected to the arduino uno v3
connector.

Signed-off-by: AJ Palmer <ajpcode@hotmail.com>

Commit:
boards: arm: nucleo_f756zg: Added SPI support

Added support for SPI 1 available from the arduino connector (plus an
extra gpio). Tested against samples/sensor/bme280.

Added warnings in documentation, pinmux and dts highlighting a potential
conflict if using SPI_1 and on-board ethernet at the same time.

Signed-off-by: AJ Palmer <ajpcode@hotmail.com>

erwango commented:
@ajpcode , please remove board.h and rebase.

galak commented on boards/arm/nucleo_f756zg/board.h:
```diff
@@ -0,0 +1,11 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __INC_BOARD_H
```
Please remove board.h

ajpcode replied:
Removed.

galak reviewed (changes requested):
Remove board.h file since it has nothing in it.

ajpcode commented:
@erwango - As requested file removed and rebased.

erwango commented on soc/arm/st_stm32/stm32f7/Kconfig.soc:
```diff
@@ -15,6 +15,9 @@ config SOC_STM32F723XE
 config SOC_STM32F746XG
 	bool "STM32F746XG"
 
+config SOC_STM32F756XG
```
Rename to STM32F756XX, since recent addition of flash an dram size in device tree files, we don't need last digit information in Kconfig.

ajpcode replied:
Changed as requested, also renamed the string on the line below and and renamed soc/arm/st_stm32/stm32f7/Kconfig.defconfig.stm32f756xg to soc/arm/st_stm32/stm32f7/Kconfig.defconfig.stm32f756xx.

ajpcode replied:
Almost forgot - also changed the Kconfig.board dependency to: "depends on SOC_STM32F756XX"

erwango reviewed (changes requested):
Last comment.
Seems good otherwise

ajpcode commented:
857bdc2 - It's been a little while since my last commit, push is just a rebase of branch on master.

erwango commented:
^^@galak

ajpcode commented:
recheck

ajpcode commented:
Rebased to master.
Added Kconfig.defconfig changes made by now merged PR #11345. (Moved from 'def_bool' to 'default').
Also, shippable should re-run :)

ajpcode commented:
Latest push:
Renamed UART_STM32_PORT_X entries in Kconfig.defconfig and pinmux.c to UART_X in line with commit id b1008cc (PR: #11234)

ajpcode commented:
@galak, @erwango  - I've rebased once again and addressed all changes I can see that have been made to the similar nucleo_f746zg board (above comments describe).

I'm not sure this shippable error is within my control (which is the following repeated):

`/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/subsys/shell/shell_uart.c:232:23: error: CONFIG_UART_SHELL_ON_DEV_NAME undeclared (first use in this function)`
```
    device_get_binding(CONFIG_UART_SHELL_ON_DEV_NAME);
                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~

erwango commented:
@apjcode, thanks for your patience and rebase efforts :-)

One thing you can do is adding `		zephyr,shell-uart = &uartX;` in your board's dts chosen node.

Though, it also reveals that shell stuff should more be controlled and we should not be forced to define it.

ajpcode commented:
@erwango - I've rebased and made the change as you suggest, I've gone with usart3 (same as the console).
